File: frontend/src/api/risk.ts
import { apiClient } from './client'

export type RiskLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'

export interface RiskFactors {
    attendance?: {
        score: number
        attendance_rate: number
        absences: number
        total_classes: number
    }
    grades?: {
        score: number
        average: number
    }
    assignments?: {
        score: number
        on_time_rate: number
        missing: number
        late: number
    }
}

export interface StudentRiskSummary {
    currentScore: number
    riskLevel: RiskLevel
    isAtRisk: boolean
    trend: 'stable' | 'increasing' | 'decreasing'
    factors: RiskFactors | null
    recommendation: string | null
    lastAssessed: string
}

export interface AtRiskStudent {
    studentId: number
    studentName: string
    riskScore: number
    mainFactor?: string
}

export interface GroupRiskDashboard {
    totalAtRisk: number
    criticalCount: number
    highCount: number
    mediumCount: number
    criticalStudents: AtRiskStudent[]
    highStudents: AtRiskStudent[]
}

export interface AttendanceStats {
    totalClasses: number
    present: number
    absent: number
    late: number
    excused: number
    attendanceRate: number
}

// Transform functions
function transformRiskSummary(data: any): StudentRiskSummary {
    return {
        currentScore: data.current_score,
        riskLevel: data.risk_level,
        isAtRisk: data.is_at_risk,
        trend: data.trend,
        factors: data.factors,
        recommendation: data.recommendation,
        lastAssessed: data.last_assessed,
    }
}

function transformDashboard(data: any): GroupRiskDashboard {
    return {
        totalAtRisk: data.total_at_risk,
        criticalCount: data.critical_count,
        highCount: data.high_count,
        mediumCount: data.medium_count,
        criticalStudents: data.critical_students?.map((s: any) => ({
            studentId: s.student_id,
            studentName: s.student_name,
            riskScore: s.risk_score,
            mainFactor: s.main_factor,
        })) || [],
        highStudents: data.high_students?.map((s: any) => ({
            studentId: s.student_id,
            studentName: s.student_name,
            riskScore: s.risk_score,
        })) || [],
    }
}

export const riskApi = {
    getMyRisk: async (groupId: number): Promise<StudentRiskSummary> => {
        const response = await apiClient.get(`/risk/my-risk/${groupId}`)
        return transformRiskSummary(response.data)
    },

    getStudentRisk: async (studentId: number, groupId: number): Promise<StudentRiskSummary> => {
        const response = await apiClient.get(`/risk/student/${studentId}/group/${groupId}`)
        return transformRiskSummary(response.data)
    },

    calculateRisk: async (studentId: number, groupId: number) => {
        const response = await apiClient.post(`/risk/calculate/${studentId}/group/${groupId}`)
        return response.data
    },

    getGroupDashboard: async (groupId: number): Promise<GroupRiskDashboard> => {
        const response = await apiClient.get(`/risk/dashboard/${groupId}`)
        return transformDashboard(response.data)
    },

    getAttendanceStats: async (studentId: number, groupId: number): Promise<AttendanceStats> => {
        const response = await apiClient.get(`/risk/attendance/${studentId}/group/${groupId}`)
        const data = response.data
        return {
            totalClasses: data.total_classes,
            present: data.present,
            absent: data.absent,
            late: data.late,
            excused: data.excused,
            attendanceRate: data.attendance_rate,
        }
    },

    recordAttendance: async (data: {
        studentId: number
        groupId: number
        classDate: string
        status: string
        notes?: string
    }) => {
        await apiClient.post('/risk/attendance', {
            student_id: data.studentId,
            group_id: data.groupId,
            class_date: data.classDate,
            status: data.status,
            notes: data.notes,
        })
    },

    recordGrade: async (data: {
        studentId: number
        groupId: number
        gradeType: string
        name: string
        grade: number
        maxGrade?: number
        weight?: number
        feedback?: string
    }) => {
        await apiClient.post('/risk/grade', {
            student_id: data.studentId,
            group_id: data.groupId,
            grade_type: data.gradeType,
            name: data.name,
            grade: data.grade,
            max_grade: data.maxGrade,
            weight: data.weight,
            feedback: data.feedback,
        })
    },

    getAtRiskStudents: async (groupId: number) => {
        const response = await apiClient.get(`/risk/at-risk-students/${groupId}`)
        return response.data
    },
}

File: backend/app/risk/schemas/risk.py
"""Pydantic schemas for Risk module."""
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field, ConfigDict


class AttendanceCreate(BaseModel):
    """Schema for recording attendance."""
    student_id: int
    group_id: int
    class_date: date
    status: str = Field(..., pattern="^(PRESENT|ABSENT|LATE|EXCUSED)$")
    notes: Optional[str] = None


class AttendanceRecord(BaseModel):
    """Schema for reading attendance."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_id: int
    group_id: int
    class_date: date
    status: str
    notes: Optional[str] = None
    created_at: datetime


class AttendanceStats(BaseModel):
    """Attendance statistics summary."""
    total_classes: int
    present: int
    absent: int
    late: int
    excused: int
    attendance_rate: float


class GradeCreate(BaseModel):
    """Schema for recording a grade."""
    student_id: int
    group_id: int
    grade_type: str = Field(..., pattern="^(EXAM|QUIZ|PARTIAL|PROJECT|HOMEWORK|PARTICIPATION|FINAL)$")
    name: str = Field(..., min_length=1, max_length=100)
    grade: float = Field(..., ge=0, le=100)
    max_grade: float = Field(default=10.0, ge=0)
    weight: float = Field(default=1.0, ge=0)
    feedback: Optional[str] = None


class GradeRecord(BaseModel):
    """Schema for reading a grade."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_id: int
    group_id: int
    grade_type: str
    name: str
    grade: float
    max_grade: float
    weight: float
    normalized_grade: float
    is_passing: bool
    feedback: Optional[str] = None
    graded_at: datetime


class RiskFactorDetail(BaseModel):
    """Detail of a single risk factor."""
    score: int
    # Additional fields depend on factor type
    data: Optional[dict] = None


class RiskAssessmentRead(BaseModel):
    """Schema for reading a risk assessment."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_id: int
    group_id: int
    risk_score: int
    risk_level: str
    attendance_score: int
    grades_score: int
    assignments_score: int
    factor_details: Optional[dict] = None
    recommendation: Optional[str] = None
    assessed_at: datetime


class StudentRiskSummary(BaseModel):
    """Summary of a student's risk status."""
    current_score: int
    risk_level: str
    is_at_risk: bool
    trend: str  # "stable", "increasing", "decreasing"
    factors: Optional[dict] = None
    recommendation: Optional[str] = None
    last_assessed: str


class AtRiskStudent(BaseModel):
    """Schema for an at-risk student in dashboard."""
    student_id: int
    student_name: str
    risk_score: int
    main_factor: Optional[str] = None


class GroupRiskDashboard(BaseModel):
    """Risk dashboard for a group."""
    total_at_risk: int
    critical_count: int
    high_count: int
    medium_count: int
    critical_students: List[AtRiskStudent] = Field(default_factory=list)
    high_students: List[AtRiskStudent] = Field(default_factory=list)
